### ISSUE TITLE
Errors while traying to fine_tune the model

### ISSUE THREAD

**Issue by VanitarNordic** [DrSleep/tensorflow-deeplab-resnet#74]
Hello,

I pass all steps and used `fine_tune.py` to fine-tune the model. I got some errors. Number of classes are 2. I used this command: `python fine_tune.py --not-restore-last`

I used these parameters:

```
IMG_MEAN = np.array((145.2201, 119.0066, 97.9356), dtype=np.float32)

BATCH_SIZE = 1
DATA_DIRECTORY = '/home/hesam/Desktop/2/train'
DATA_LIST_PATH = 'data/train.txt'
IGNORE_LABEL = 255
INPUT_SIZE = '321,321'
LEARNING_RATE = 1e-4
NUM_CLASSES = 2
NUM_STEPS = 20000
RANDOM_SEED = 1234
RESTORE_FROM = 'data/deeplab_resnet.ckpt'
SAVE_NUM_IMAGES = 2
SAVE_PRED_EVERY = 100
SNAPSHOT_DIR = 'data'

```

and

```
# colour map
label_colours = [(0,0,0),(131,0,5)]
```

**The errors:** 

```
(tensorflow) hesam@hesam-MS-7994:~/Desktop/tensorflow-deeplab-resnet-master$ python fine_tune.py --not-restore-last
Couldn't import dot_parser, loading of dot files will not be possible.
2017-04-26 11:26:40.077467: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.1 instructions, but these are available on your machine and could speed up CPU computations.
2017-04-26 11:26:40.077492: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.
2017-04-26 11:26:40.077496: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.
2017-04-26 11:26:40.077499: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX2 instructions, but these are available on your machine and could speed up CPU computations.
2017-04-26 11:26:40.077502: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use FMA instructions, but these are available on your machine and could speed up CPU computations.
2017-04-26 11:26:40.195856: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:901] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2017-04-26 11:26:40.196757: I tensorflow/core/common_runtime/gpu/gpu_device.cc:887] Found device 0 with properties: 
name: GeForce GTX 1060 6GB
major: 6 minor: 1 memoryClockRate (GHz) 1.759
pciBusID 0000:01:00.0
Total memory: 5.93GiB
Free memory: 5.54GiB
2017-04-26 11:26:40.196771: I tensorflow/core/common_runtime/gpu/gpu_device.cc:908] DMA: 0 
2017-04-26 11:26:40.196775: I tensorflow/core/common_runtime/gpu/gpu_device.cc:918] 0:   Y 
2017-04-26 11:26:40.196780: I tensorflow/core/common_runtime/gpu/gpu_device.cc:977] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX 1060 6GB, pci bus id: 0000:01:00.0)
Restored model parameters from data/deeplab_resnet.ckpt
2017-04-26 11:26:46.111049: W tensorflow/core/framework/op_kernel.cc:1152] Not found: /home/hesam/Desktop/2/train/labels/0039.png
2017-04-26 11:26:46.112367: W tensorflow/core/framework/op_kernel.cc:1152] Not found: /home/hesam/Desktop/2/train/labels/0039.png
	 [[Node: create_inputs/ReadFile_1 = ReadFile[_device="/job:localhost/replica:0/task:0/cpu:0"](create_inputs/input_producer/Gather_1)]]
2017-04-26 11:26:46.222843: W tensorflow/core/framework/op_kernel.cc:1152] Out of range: FIFOQueue '_1_create_inputs/batch/fifo_queue' is closed and has insufficient elements (requested 1, current size 0)
	 [[Node: create_inputs/batch = QueueDequeueManyV2[component_types=[DT_FLOAT, DT_UINT8], timeout_ms=-1, _device="/job:localhost/replica:0/task:0/cpu:0"](create_inputs/batch/fifo_queue, create_inputs/batch/n)]]
2017-04-26 11:26:46.223115: W tensorflow/core/framework/op_kernel.cc:1152] Out of range: FIFOQueue '_1_create_inputs/batch/fifo_queue' is closed and has insufficient elements (requested 1, current size 0)
	 [[Node: create_inputs/batch = QueueDequeueManyV2[component_types=[DT_FLOAT, DT_UINT8], timeout_ms=-1, _device="/job:localhost/replica:0/task:0/cpu:0"](create_inputs/batch/fifo_queue, create_inputs/batch/n)]]
Traceback (most recent call last):
  File "fine_tune.py", line 207, in <module>
    main()
  File "fine_tune.py", line 196, in main
    loss_value, images, labels, preds, summary, _ = sess.run([reduced_loss, image_batch, label_batch, pred, total_summary, optim])
  File "/home/hesam/Desktop/tensorflow/local/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 778, in run
    run_metadata_ptr)
  File "/home/hesam/Desktop/tensorflow/local/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 982, in _run
    feed_dict_string, options, run_metadata)
  File "/home/hesam/Desktop/tensorflow/local/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 1032, in _do_run
    target_list, options, run_metadata)
  File "/home/hesam/Desktop/tensorflow/local/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 1052, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.OutOfRangeError: FIFOQueue '_1_create_inputs/batch/fifo_queue' is closed and has insufficient elements (requested 1, current size 0)
	 [[Node: create_inputs/batch = QueueDequeueManyV2[component_types=[DT_FLOAT, DT_UINT8], timeout_ms=-1, _device="/job:localhost/replica:0/task:0/cpu:0"](create_inputs/batch/fifo_queue, create_inputs/batch/n)]]

Caused by op u'create_inputs/batch', defined at:
  File "fine_tune.py", line 207, in <module>
    main()
  File "fine_tune.py", line 125, in main
    image_batch, label_batch = reader.dequeue(args.batch_size)
  File "/home/hesam/Desktop/tensorflow-deeplab-resnet-master/deeplab_resnet/image_reader.py", line 179, in dequeue
    num_elements)
  File "/home/hesam/Desktop/tensorflow/local/lib/python2.7/site-packages/tensorflow/python/training/input.py", line 917, in batch
    name=name)
  File "/home/hesam/Desktop/tensorflow/local/lib/python2.7/site-packages/tensorflow/python/training/input.py", line 712, in _batch
    dequeued = queue.dequeue_many(batch_size, name=name)
  File "/home/hesam/Desktop/tensorflow/local/lib/python2.7/site-packages/tensorflow/python/ops/data_flow_ops.py", line 458, in dequeue_many
    self._queue_ref, n=n, component_types=self._dtypes, name=name)
  File "/home/hesam/Desktop/tensorflow/local/lib/python2.7/site-packages/tensorflow/python/ops/gen_data_flow_ops.py", line 1328, in _queue_dequeue_many_v2
    timeout_ms=timeout_ms, name=name)
  File "/home/hesam/Desktop/tensorflow/local/lib/python2.7/site-packages/tensorflow/python/framework/op_def_library.py", line 768, in apply_op
    op_def=op_def)
  File "/home/hesam/Desktop/tensorflow/local/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 2336, in create_op
    original_op=self._default_original_op, op_def=op_def)
  File "/home/hesam/Desktop/tensorflow/local/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 1228, in __init__
    self._traceback = _extract_stack()

OutOfRangeError (see above for traceback): FIFOQueue '_1_create_inputs/batch/fifo_queue' is closed and has insufficient elements (requested 1, current size 0)
	 [[Node: create_inputs/batch = QueueDequeueManyV2[component_types=[DT_FLOAT, DT_UINT8], timeout_ms=-1, _device="/job:localhost/replica:0/task:0/cpu:0"](create_inputs/batch/fifo_queue, create_inputs/batch/n)]]


```

**Comment by DrSleep** [comment]
Looks like the files are not presented:

> 2017-04-26 11:26:46.111049: W tensorflow/core/framework/op_kernel.cc:1152] Not found: /home/hesam/Desktop/2/train/labels/0039.png
> 2017-04-26 11:26:46.112367: W tensorflow/core/framework/op_kernel.cc:1152] Not found: /home/hesam/Desktop/2/train/labels/0039.png

**Comment by VanitarNordic** [comment]
Strange, files are exist in the path. This is what inside `train.txt`:


```
/images/0000.jpg /labels/0000.png
/images/0002.jpg /labels/0002.png
/images/0005.jpg /labels/0005.png
/images/0006.jpg /labels/0006.png
/images/0007.jpg /labels/0007.png
/images/0008.jpg /labels/0008.png
/images/0009.jpg /labels/0009.png
/images/0010.jpg /labels/0010.png
/images/0011.jpg /labels/0011.png
/images/0012.jpg /labels/0012.png
/images/0013.jpg /labels/0013.png
.
.
.
.

```


The only thing is that I made this file by writing a software in Windows. I mean `train.txt` is generated in windows then I copied to Ubuntu drive to use it. I don't see any problem inside when I open the file, but is there any difference that I don't know?!

**Comment by VanitarNordic** [comment]
Okay, I fixed the above mentioned issue by a command named: `does2unix` to convert windows generated `txt` files to Linux `txt` file.

But I got some new errors:

```
python fine_tune.py --not-restore-last
Couldn't import dot_parser, loading of dot files will not be possible.
2017-04-26 14:29:01.634345: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.1 instructions, but these are available on your machine and could speed up CPU computations.
2017-04-26 14:29:01.634366: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.
2017-04-26 14:29:01.634373: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.
2017-04-26 14:29:01.634377: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX2 instructions, but these are available on your machine and could speed up CPU computations.
2017-04-26 14:29:01.634383: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use FMA instructions, but these are available on your machine and could speed up CPU computations.
2017-04-26 14:29:01.753617: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:901] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2017-04-26 14:29:01.753926: I tensorflow/core/common_runtime/gpu/gpu_device.cc:887] Found device 0 with properties: 
name: GeForce GTX 1060 6GB
major: 6 minor: 1 memoryClockRate (GHz) 1.759
pciBusID 0000:01:00.0
Total memory: 5.93GiB
Free memory: 5.53GiB
2017-04-26 14:29:01.753939: I tensorflow/core/common_runtime/gpu/gpu_device.cc:908] DMA: 0 
2017-04-26 14:29:01.753944: I tensorflow/core/common_runtime/gpu/gpu_device.cc:918] 0:   Y 
2017-04-26 14:29:01.753952: I tensorflow/core/common_runtime/gpu/gpu_device.cc:977] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX 1060 6GB, pci bus id: 0000:01:00.0)
Restored model parameters from data/deeplab_resnet.ckpt
Traceback (most recent call last):
  File "/home/hesam/Desktop/tensorflow/local/lib/python2.7/site-packages/tensorflow/python/ops/script_ops.py", line 82, in __call__
    ret = func(*args)
  File "/home/hesam/Desktop/tensorflow-deeplab-resnet-master/deeplab_resnet/utils.py", line 21, in decode_labels
    assert(n >= num_images), 'Batch size %d should be greater or equal than number of images to save %d.' % (n, num_images)
AssertionErrorTraceback (most recent call last):
  File "/home/hesam/Desktop/tensorflow/local/lib/python2.7/site-packages/tensorflow/python/ops/script_ops.py", line 82, in __call__
:     Batch size 1 should be greater or equal than number of images to save 2.
ret = func(*args)
  File "/home/hesam/Desktop/tensorflow-deeplab-resnet-master/deeplab_resnet/utils.py", line 66, in inv_preprocess
    assert(n >= num_images), 'Batch size %d should be greater or equal than number of images to save %d.' % (n, num_images)
AssertionError: Batch size 1 should be greater or equal than number of images to save 2.
2017-04-26 14:29:07.581577: W tensorflow/core/framework/op_kernel.cc:1152] Internal: Failed to run py callback pyfunc_0: see error log.
2017-04-26 14:29:07.581577: W tensorflow/core/framework/op_kernel.cc:1152] Internal: Failed to run py callback pyfunc_1: see error log.
2017-04-26 14:29:07.581748: W tensorflow/core/framework/op_kernel.cc:1152] Internal: Failed to run py callback pyfunc_0: see error log.
	 [[Node: PyFunc = PyFunc[Tin=[DT_FLOAT, DT_INT32, DT_FLOAT], Tout=[DT_UINT8], token="pyfunc_0", _device="/job:localhost/replica:0/task:0/cpu:0"](create_inputs/batch, PyFunc/input_1, PyFunc/input_2)]]
2017-04-26 14:29:07.581777: W tensorflow/core/framework/op_kernel.cc:1152] Internal: Failed to run py callback pyfunc_0: see error log.
	 [[Node: PyFunc = PyFunc[Tin=[DT_FLOAT, DT_INT32, DT_FLOAT], Tout=[DT_UINT8], token="pyfunc_0", _device="/job:localhost/replica:0/task:0/cpu:0"](create_inputs/batch, PyFunc/input_1, PyFunc/input_2)]]
2017-04-26 14:29:10.937077: W tensorflow/core/common_runtime/bfc_allocator.cc:217] Allocator (GPU_0_bfc) ran out of memory trying to allocate 2.71GiB. The caller indicates that this is not a failure, but may mean that there could be performance gains if more memory is available.
2017-04-26 14:29:10.954844: W tensorflow/core/common_runtime/bfc_allocator.cc:217] Allocator (GPU_0_bfc) ran out of memory trying to allocate 2.54GiB. The caller indicates that this is not a failure, but may mean that there could be performance gains if more memory is available.
2017-04-26 14:29:10.968999: W tensorflow/core/framework/op_kernel.cc:1152] Internal: Failed to run py callback pyfunc_0: see error log.
	 [[Node: PyFunc = PyFunc[Tin=[DT_FLOAT, DT_INT32, DT_FLOAT], Tout=[DT_UINT8], token="pyfunc_0", _device="/job:localhost/replica:0/task:0/cpu:0"](create_inputs/batch, PyFunc/input_1, PyFunc/input_2)]]
2017-04-26 14:29:10.969069: W tensorflow/core/framework/op_kernel.cc:1152] Internal: Failed to run py callback pyfunc_0: see error log.
	 [[Node: PyFunc = PyFunc[Tin=[DT_FLOAT, DT_INT32, DT_FLOAT], Tout=[DT_UINT8], token="pyfunc_0", _device="/job:localhost/replica:0/task:0/cpu:0"](create_inputs/batch, PyFunc/input_1, PyFunc/input_2)]]
Traceback (most recent call last):
  File "fine_tune.py", line 207, in <module>
    main()
  File "fine_tune.py", line 196, in main
    loss_value, images, labels, preds, summary, _ = sess.run([reduced_loss, image_batch, label_batch, pred, total_summary, optim])
  File "/home/hesam/Desktop/tensorflow/local/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 778, in run
    run_metadata_ptr)
  File "/home/hesam/Desktop/tensorflow/local/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 982, in _run
    feed_dict_string, options, run_metadata)
  File "/home/hesam/Desktop/tensorflow/local/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 1032, in _do_run
    target_list, options, run_metadata)
  File "/home/hesam/Desktop/tensorflow/local/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 1052, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.InternalError: Failed to run py callback pyfunc_0: see error log.
	 [[Node: PyFunc = PyFunc[Tin=[DT_FLOAT, DT_INT32, DT_FLOAT], Tout=[DT_UINT8], token="pyfunc_0", _device="/job:localhost/replica:0/task:0/cpu:0"](create_inputs/batch, PyFunc/input_1, PyFunc/input_2)]]

Caused by op u'PyFunc', defined at:
  File "fine_tune.py", line 207, in <module>
    main()
  File "fine_tune.py", line 158, in main
    images_summary = tf.py_func(inv_preprocess, [image_batch, args.save_num_images, IMG_MEAN], tf.uint8)
  File "/home/hesam/Desktop/tensorflow/local/lib/python2.7/site-packages/tensorflow/python/ops/script_ops.py", line 189, in py_func
    input=inp, token=token, Tout=Tout, name=name)
  File "/home/hesam/Desktop/tensorflow/local/lib/python2.7/site-packages/tensorflow/python/ops/gen_script_ops.py", line 40, in _py_func
    name=name)
  File "/home/hesam/Desktop/tensorflow/local/lib/python2.7/site-packages/tensorflow/python/framework/op_def_library.py", line 768, in apply_op
    op_def=op_def)
  File "/home/hesam/Desktop/tensorflow/local/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 2336, in create_op
    original_op=self._default_original_op, op_def=op_def)
  File "/home/hesam/Desktop/tensorflow/local/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 1228, in __init__
    self._traceback = _extract_stack()

InternalError (see above for traceback): Failed to run py callback pyfunc_0: see error log.
	 [[Node: PyFunc = PyFunc[Tin=[DT_FLOAT, DT_INT32, DT_FLOAT], Tout=[DT_UINT8], token="pyfunc_0", _device="/job:localhost/replica:0/task:0/cpu:0"](create_inputs/batch, PyFunc/input_1, PyFunc/input_2)]]

```

**Comment by VanitarNordic** [comment]
Alright, the problem solved. I had to increase the batch size from 1 to 2. because as error says it should be equal or bigger than number of images in `utils.py`.

as the model trains, it creates 3 different formats in the target directory: for example for the 100th snapshot:

```
model.ckpt-100.data-00000-of-00001
model.ckpt-100.index
model.ckpt-100.meta
```

But now I faced a problem with `evaluate.py` . if I could not solve, I'll post the issue here.

**Comment by VanitarNordic** [comment]
Alright, the problem solved. I had to increase the batch size from 1 to 2. because as error says it should be equal or bigger than number of images in `utils.py`.

as the model trains, it creates 3 different formats in the target directory: for example for the 100th snapshot:

```
model.ckpt-100.data-00000-of-00001
model.ckpt-100.index
model.ckpt-100.meta
```

But now I faced a problem with `evaluate.py` . if I could not solve, I'll post the issue here. The question is which file should be used for evaluation? I suppose the first one, yes?

**Comment by VanitarNordic** [comment]
I renamed the first file to mode.ckpt and tried to evaluate the model on a validation data-set with 24 images (`evaluate.py`). for the IMG_MEAN I used the values for the training because before the modification, you had put same IMG_MEAN for all python files.

Parameters:

```
DATA_DIRECTORY = '/home/hesam/Desktop/2/val'
DATA_LIST_PATH = 'data/val.txt'
IGNORE_LABEL = 255
NUM_CLASSES = 2
NUM_STEPS = 24 # Number of images in the validation set.
RESTORE_FROM = 'data/model.ckpt'
```

**Error:**

```
DataLossError (see above for traceback): Unable to open table file data/model.ckpt: Data loss: not an sstable (bad magic number): perhaps your file is in a different file format and you need to use a different restore operator?
	 [[Node: save/RestoreV2_65 = RestoreV2[dtypes=[DT_FLOAT], _device="/job:localhost/replica:0/task:0/cpu:0"](_recv_save/Const_0, save/RestoreV2_65/tensor_names, save/RestoreV2_65/shape_and_slices)]]
	 [[Node: save/RestoreV2_149/_405 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/gpu:0", send_device="/job:localhost/replica:0/task:0/cpu:0", send_device_incarnation=1, tensor_name="edge_2723_save/RestoreV2_149", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/gpu:0"]()]]


```

**Comment by DrSleep** [comment]
No need to rename. Just use the path 'model.ckpt-100': it will fetch up everything

**Comment by VanitarNordic** [comment]
Thank you.

I did for example for 1800th snapshot:
`
RESTORE_FROM = 'data/model.ckpt-1800'`

But I got this error (a different error):

```
NotFoundError (see above for traceback): Key fc1_voc12_c0/biases not found in checkpoint
	 [[Node: save/RestoreV2_417 = RestoreV2[dtypes=[DT_FLOAT], _device="/job:localhost/replica:0/task:0/cpu:0"](_recv_save/Const_0, save/RestoreV2_417/tensor_names, save/RestoreV2_417/shape_and_slices)]]
	 [[Node: save/RestoreV2_17/_1005 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/gpu:0", send_device="/job:localhost/replica:0/task:0/cpu:0", send_device_incarnation=1, tensor_name="edge_387_save/RestoreV2_17", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/gpu:0"]()]]``

**Comment by DrSleep** [comment]
https://github.com/DrSleep/tensorflow-deeplab-resnet/issues/75

TL;DR: there was a bug, which is now fixed. Please clone the repository again

**Comment by VanitarNordic** [comment]
Thank you, now the error has solved, validation give this output as a result:

```
step 0
Mean IoU: 0.500
```

But you know 50% is quite low for a fine-tunned model. it is lower than what is trained from scratch for a big dataset such as VOC. What's your suggestion? is it because that I used the IMG_MEAN of training images?

**Comment by DrSleep** [comment]
Well, there may be several suggestions as you would definitely need to play with hyperparameters: I would say that learning rate and batch size are the most crucial ones. I would also suggest to track how the validation score changes over the training process and how the predictions change. Also take a look at the performance on the training set: if it is significantly higher, then weight decay should be tuned, as well.

**Comment by VanitarNordic** [comment]
Well, I can say the IOU from the 100th snapshot does not make any difference with 2000th. both are 0.5. I don't know why.
This picture shows the errors. is there possibility to change the solver? I mean SGD, Adam, AdaDelta and so on?

![screenshot from 2017-04-26 19-47-53](https://cloud.githubusercontent.com/assets/18719591/25442189/9b7e7792-2ab9-11e7-96ac-ebbf9dd9392a.png)

**Comment by VanitarNordic** [comment]
Also, you have not mentioned about this in custom training, but there is a `num_classes` parameter inside `utils.py` , which was set to 21. 

```
# colour map
label_colours = [(0,0,0),(131,0,5)]
               
def decode_labels(mask, num_images=1, num_classes=21):
```

this should not be changed? Although I can say I changed it and no difference!

**Comment by VanitarNordic** [comment]
Also, I should say that I ran the evaluation on the **training** images also and the IOU is 0.5 again which it MUST not be! 
Definitely something is wrong somewhere!

**Comment by VanitarNordic** [comment]
I can send you my dataset by Email to check it yourself. it is only 10MB annotated images for 2 classes. 
I got good results on this small data-set with the fine-tuned FCN-8s. So I want to compare the identical dataset with two different models, and see the results.
DeepLabV2 is at least 15% better on VOC on training from scratch!. 
So I assume I must get better results! not lower.

**Comment by DrSleep** [comment]
Could you check the tensorboard images?  

> Also, you have not mentioned about this in the custom training, but there is a num_classes parameter inside utils.py , which was set to 21.
The decode labels function is only called from training, inference or evaluation scripts, all of which have the number of classes parameter.

Sure, you can send it to me (you can find my email in the profile). I would take a look at it when I have time.

**Comment by VanitarNordic** [comment]
> Could you check the tensorboard images?
You mean checking the graphs?

Have you tested the repo on different number of classes yourself?

**Comment by DrSleep** [comment]
I mean the actual image summary.

I trained on a small subset of ADE some time ago; I did not test the performance in mIoU, just the results - they looked fine.

**Comment by VanitarNordic** [comment]
> I mean the actual image summary.

Where can I find it? I mean does it saved to a specific path?

**Comment by DrSleep** [comment]
The 'images' tab in tensorboard
https://github.com/DrSleep/tensorflow-deeplab-resnet/raw/master/images/summary.png

**Comment by VanitarNordic** [comment]
Here it is. it seems the model is confused and trained very bad since the IOU on training is also 0.5!

![screenshot from 2017-04-27 11-31-43](https://cloud.githubusercontent.com/assets/18719591/25471721/97fab142-2b3d-11e7-8953-ca83a94f1a6e.png)

**Comment by DrSleep** [comment]
Looks like your ground truth masks are not converted correctly.

**Comment by VanitarNordic** [comment]
> Looks like your ground truth masks are not converted correctly.

I've created the color map correctly. The only thing is that my dataset contains some background-only images also which I don't think would matter since I 've tested it before with a different model. I send it to you now.

**Comment by DrSleep** [comment]
Your label masks are 3-d tensors: i.e., there are values like 0, 128, 192, 224. Instead, each 3d vector should be converted to a number. You need to save your annotations without colourmap, as also written in the DeepLab FAQ: http://liangchiehchen.com/projects/DeepLab_FAQ.html

**Comment by VanitarNordic** [comment]
Sorry I don't get the point since my labels are like VOC, which each label corresponds to a class, for example red is Airplane with some RGB value. otherwise it gonna be grayscale isn't it?

**Comment by DrSleep** [comment]
Please take a look here: https://github.com/martinkersner/train-DeepLab#data-conversions (the second paragraph).
Also this function should be useful to perform the conversion: https://github.com/martinkersner/train-DeepLab/blob/master/utils.py#L91

**Comment by VanitarNordic** [comment]
Okay. Thank you. 

The author has introduced a matlab script also: `SavePngAsRawPng.m` to make the conversion but it seems it just read PNG images and writes them somewhere else with no specific operation of them.

**Comment by VanitarNordic** [comment]
if you just tell me what is the term for this conversion, I will write the code myself in Matlab or something. 
I mean should I convert PNG32 to PNG8 or similar?

**Comment by VanitarNordic** [comment]
I converted all PNG labels to PNG-8, but still no improvement in results :-( both training and validation shows IOU of 0.5 all the times.

**Comment by VanitarNordic** [comment]
Alright, I solved the case finally. The term for this conversion is indexing the image. I'll make a repository to help other people who may stuck at this stage.

Now the IOU on validation is about 0.917. But I have a question on inference. inference results show a back image. is there a better script to show the segmentation on the original image itself?

**Comment by DrSleep** [comment]
You are welcome to extend the repository. All the functional is described in README. Please refer to it.

**Comment by VanitarNordic** [comment]
I see the model is less sensitive to small objects, Do you have a suggestion to solve it?

**Comment by VanitarNordic** [comment]
@DrSleep 

Please have a look at this image (second images). it seems the model misunderstood the ignore_label as a class. Why this happened?

![screenshot from 2017-04-28 17-45-54](https://cloud.githubusercontent.com/assets/18719591/25530247/e1cb7966-2c3a-11e7-8da4-c0a920110f8b.png)

**Comment by VanitarNordic** [comment]
@DrSleep 

Okay. I reached an IOU on my own dataset as 0.93, but you know the segmentation is not as good as FCN-8s which is a weaker model in comparison with DeepLabV2. any suggestion is really appreciated.
